torchok/models/heads/detection/yolov3.py

```diff
@@ -0,0 +1,231 @@
+
+        Returns:
+            list[tuple[Tensor, Tensor]]: Each item in result_list is 2-tuple.
+                The first item is an (n, 5) tensor, where 5 represent
+                (tl_x, tl_y, br_x, br_y, score) and the score between 0 and 1.
+                The shape of the second tensor in the tuple is (n,), and
+                each element represents the class label of the corresponding
+                box.
+        """
+        cfg = self.test_cfg
+
+        num_imgs = len(pred_maps[0])
+        featmap_sizes = [pred_map.shape[-2:] for pred_map in pred_maps]
+
+        mlvl_anchors = self.prior_generator.grid_priors(
+            featmap_sizes, device=pred_maps[0].device)
+        flatten_preds = []
+        flatten_strides = []
+        for pred, stride in zip(pred_maps, self.featmap_strides):
+            pred = pred.permute(0, 2, 3, 1).reshape(num_imgs, -1, self.num_attrib)
+            pred[..., :2].sigmoid_()
+            flatten_preds.append(pred)
+            flatten_strides.append(pred.new_tensor(stride).expand(pred.size(1)))
+
+        flatten_preds = torch.cat(flatten_preds, dim=1)
+        flatten_bbox_preds = flatten_preds[..., :4]
+        flatten_objectness = flatten_preds[..., 4].sigmoid()
+        flatten_cls_scores = flatten_preds[..., 5:].sigmoid()
+        flatten_anchors = torch.cat(mlvl_anchors)
+        flatten_strides = torch.cat(flatten_strides)
+        flatten_bboxes = self.bbox_coder.decode(flatten_anchors,
+                                                flatten_bbox_preds,
+                                                flatten_strides.unsqueeze(-1))
+
+        if with_nms and (flatten_objectness.size(0) == 0):
+            return torch.zeros((0, 5)), torch.zeros((0,))
+
+        padding = flatten_bboxes.new_zeros(num_imgs, flatten_bboxes.shape[1], 1)
+        flatten_cls_scores = torch.cat([flatten_cls_scores, padding], dim=-1)
+
+        det_results = []
+        for (bboxes, scores, objectness) in zip(flatten_bboxes,
+                                                flatten_cls_scores,
+                                                flatten_objectness):
+            # Filtering out all predictions with conf < conf_thr
+            conf_thr = cfg.get('conf_thr', -1)
+            if conf_thr > 0:
+                conf_inds = objectness >= conf_thr
+                bboxes = bboxes[conf_inds, :]
+                scores = scores[conf_inds, :]
+                objectness = objectness[conf_inds]
+
+            det_bboxes, det_labels = multiclass_nms(
+                bboxes,
+                scores,
+                cfg['score_thr'],
+                dict(cfg['nms']),
+                cfg['max_per_img'],
+                score_factors=objectness)
+            det_results.append(dict(bboxes=det_bboxes, labels=det_labels))
+        return det_results
+
+    @force_fp32(apply_to=('pred_maps',))
+    def loss(self, joint_loss: JointLoss, pred_maps: List[Tensor], gt_bboxes: List[Tensor], gt_labels: List[Tensor],
```

Comment:
In MMDetection loss and get_bboxes method has already writen. Why we need to overwrite this method here?

Reply:
Because we have JointLoss that cannot be adapted otherwise from mmdet